A80-pid.md

```diff
@@ -0,0 +1,312 @@
+	// * If multiple updates happened during a single WeightUpdatePeriod, the actual weights are not applied,
+	// but PID controller keep growing the weight and it may easily pass the balancing point.
+	if time.Since(lastApplied) < conf.WeightUpdatePeriod {
+		return -1
+	}
+
+  // use value calculated in the onEDFSchedulerUpdate method
+	meanUtilization = data.meanUtilization
+
+  // call PID controlelr to get the value of the control signal.
+	controlSignal = data.pidController.update({
+		referenceSignal:  meanUtilization,
+		actualSignal:     utilization,
+		samplingInterval: time.Since(lastApplied),
+	})
+
+	// Normalize the signal.
+	// If meanUtilization ~= 0 the signal will be ~= 0 as well, and convergence will becoma painfully slow.
+	// If, meanUtilization >> 1 the signal may become very high, which could lead to oscillations.
+	if meanUtilization > 0 {
+		controlSignal *= 1 / meanUtilization
+	}
+
+  lastAppliedWeight = data.lastAppliedWeightPerSubchannel[subchannelID]
+
+  // Use controlSignal to adjust the weight.
+  // First caclulate multiplier that will be used to determine how much weight should be changed.
+  // The higher is the absolute value of the controlSignal the more we need to adjust the weight.
+	if controlSignal >= 0 {
+    // in this case mult should belong to [1,inf) interval, so we will be increasing the weight.
+		mult = 1.0 + controlSignal
+	} else {
+    // in this case mult should belong to (0, 1) interval, so we will be decreasing the weight.
+		mult = -1.0 / (controlSignal - 1.0)
+	}
+	weight = lastAppliedWeight * mult
+
+  // Clamp weight
+	if weight > conf.MaxWeight {
+		weight = conf.MaxWeight
+	}
+	if weight < conf.MinWeight {
+		weight = conf.MinWeight
+	}
+
+  // Save resulting utilization and weight.
+	data.utilizationPerSubchannel[subchannelId] = utilization
+  data.lastAppliedWeightPerSubchannel[subchannelID] = weight
+
+  return weight
+}
+
+func onEDFSchedulerUpdate(data callbackData) {
+  // Calculate mean utilization across all subchannels
+  totalUtilization = 0
+  count = len(data.utilizationPerSubchannel)
+  for _, utilization in data.utilizationPerSubchannel {
+    totalUtilization += utilization
+  }
+  data.meanUtilization = totalUtilization / count
+}
+```
+
+### Dealing with Oscillations
+
+One of the main challenges with the pid balancer is the potential for oscillations. Several factors influence this likelihood:
+
+1. **Propagation Delay of Load Reports:**
+  * **Direct Load Reporting**: Here, the delay depends on the request frequency and the `WeightUpdatePeriod` setting. Typically, with the default `WeightUpdatePeriod` of 1 second, propagation is very fast, making this the preferred option when using the pid balancer.
+  * **OOB Load Reporting**: Users can control the delay by adjusting the `OobReportingPeriod` setting. While the delay is usually larger compared to direct reporting, achieving perfect convergence with OOB reporting is still possible on workloads with stable loads.
```

Comment:
I'm still a bit fuzzy as to why the PID balancer doesn't automatically imply subsetting too. They seem tightly coupled to me.

In my mental model, OOB is a bit at odds with subsetting. High level, I think of the primary goal of subsetting as ensuring a client reserves a slice of the backends capacity that is commensurate to its offered load. That is, if we subset correctly a client's subset window will be hot. Thus, the need for OOB seems less relevant. Put differently, if we need OOB load reporting, we are likely sub optimally sizing the slice.

Reply:
PID might be useful without subsetting to correct load imbalance that is due to some other factors, like heterogeneous requests, but I agree that it is mostly useful in combination with subsetting.

I was thinking about disabling OOB with PID, I'll ask maintainers about that and do what they prefer.